### PR TITLE
[CodeCamp2023-287] Add `DVCLiveVisBackend`

### PR DESCRIPTION
**Pull request by RangeKing**
## Motivation

Add `DVCLiveVisBackend`

**Note:**
The latest vesion of `dvclive` in Python 3.7 is `0.9.0` which doesn't support `log_param` and `log_params`.
Therefore, Python version should be 3.8 or higher version to use `DVCLiveVisBackend`.

### CODE REVIEW

File: mmengine/visualization/vis_backend.py [hunk 1]
```diff
@@ -1130,3 +1131,174 @@ def close(self) -> None:
...
+                   **kwargs) -> None:
+        """Record the scalar data to dvclive.
+
+        Args:
+            name (str): The scalar identifier.
+            value (int, float, torch.Tensor, np.ndarray): Value to save.
+            step (int): Useless parameter. Dvclive does not
+                need this parameter. Defaults to 0.
+        """
+        self._dvclive.log_param(name, value)
```
Comment (zhouzaida) [inline mmengine/visualization/vis_backend.py:1258]: Calling the `log_metric` than `log_param` should be proper.

File: mmengine/visualization/vis_backend.py [hunk 2]
```diff
@@ -1130,3 +1131,174 @@ def close(self) -> None:
...
+    Examples:
+        >>> from mmengine.visualization import DVCLiveVisBackend
+        >>> import numpy as np
+        >>> dvclive_vis_backend = DVCLiveVisBackend(save_dir='temp_dir')
+        >>> img=np.random.randint(0, 256, size=(10, 10, 3))
+        >>> dvclive_vis_backend.add_image('img', img)
+        >>> dvclive_vis_backend.add_scalar('mAP', 0.6)
+        >>> dvclive_vis_backend.add_scalars({'loss': 0.1, 'acc': 0.8})
+        >>> cfg = Config(dict(a=1, b=dict(b1=[0, 1])))
+        >>> dvclive_vis_backend.add_config(cfg)
```
Comment (zhouzaida) [inline mmengine/visualization/vis_backend.py:1149]:
```suggestion
        >>> dvclive_vis_backend.add_config(cfg)
        
    Note:
        `New in version 0.8.5.`
```

File: mmengine/visualization/vis_backend.py [hunk 3]
```diff
@@ -1130,3 +1131,174 @@ def close(self) -> None:
...
+        """
+        self._dvclive.log_params(scalar_dict)
+
+    def close(self) -> None:
+        """close an opened dvclive object."""
+        if not hasattr(self, '_dvclive'):
+            return
+
+        file_paths = dict()
+        for filename in scandir(self.cfg.work_dir, self._artifact_suffix,
```
Comment (zhouzaida) [inline mmengine/visualization/vis_backend.py:1284]: If `add_config` is not called, `self.cfg` will throw an error.

Comment (RangeKing) [reply mmengine/visualization/vis_backend.py:1289]: I just copied the codes from `MLflowVisBackend`. How can I solve this?

Comment (RangeKing) [reply mmengine/visualization/vis_backend.py:1289]: > I just copied the codes from `MLflowVisBackend`. How can I solve this problem?

**Comment by zhouzaida** [comment]
Could you help update the usage document in `docs/zh_cn/common_usage/visualize_training_log.md` and `docs/en/common_usage/visualize_training_log.md`?

File: mmengine/visualization/vis_backend.py (hunk 3 above)
Comment (zhouzaida) [reply mmengine/visualization/vis_backend.py:1289]: You can refer to https://github.com/open-mmlab/mmengine/blob/5671b53bc575f8165b1a296e6e9cf007c7de08de/mmengine/visualization/vis_backend.py#L979

File: mmengine/visualization/vis_backend.py [hunk 4]
```diff
@@ -1130,3 +1131,190 @@ def close(self) -> None:
...
+        if self._init_kwargs is None:
+            self._init_kwargs = {
+                'dir': self._save_dir,
+                'save_dvc_exp': True,
+                'cache_images': True,
+            }
+        else:
+            self._init_kwargs.setdefault('dir', self._save_dir)
+            self._init_kwargs.setdefault('save_dvc_exp', True)
+            self._init_kwargs.setdefault('cache_images', True)
```
Comment (zhouzaida) [inline mmengine/visualization/vis_backend.py:1201]:
```suggestion
        if self._init_kwargs is None:
            self._init_kwargs = {}
        self._init_kwargs.setdefault('dir', self._save_dir)
        self._init_kwargs.setdefault('save_dvc_exp', True)
        self._init_kwargs.setdefault('cache_images', True)
```

File: mmengine/visualization/vis_backend.py [hunk 5]
```diff
@@ -1130,3 +1131,190 @@ def close(self) -> None:
...
+
+        Args:
+            scalar_dict (dict): Key-value pair storing the tag and
+                corresponding values.
+            step (int): Useless parameter. Dvclive does not
+                need this parameter. Defaults to 0.
+            file_path (str, optional): Useless parameter. Just for
+                interface unification. Defaults to None.
+        """
+        self._dvclive.log_params(self._to_dvc_paramlike(scalar_dict))
```
Comment (zhouzaida) [inline mmengine/visualization/vis_backend.py:1281]:
```suggestion
        for key, value in scalar_dict.items():
            self.add_scalar(key, value, step, **kwargs)
```

**Comment by RangeKing** [comment]
> Could you help update the usage document in `docs/zh_cn/common_usage/visualize_training_log.md` and `docs/en/common_usage/visualize_training_log.md`?

With pleasure.

**Comment by zhouzaida** [comment]
> > Could you help update the usage document in `docs/zh_cn/common_usage/visualize_training_log.md` and `docs/en/common_usage/visualize_training_log.md`?
> 
> With pleasure. Now there is a problem. DVC works best in a Git repository and most users run experiments in a OpenMMLab repo. But `work_dirs/` is in the `.gitignore` file. Therefore DVC will give a warning `WARNING:dvclive:Error in cache: bad DVC file name 'work_dirs\xxx.dvc' is git-ignored`.

How about opening an issue on dvclive?

**Comment by RangeKing** [comment]
@zhouzaida I wrote a note in document. Recommend not to set `work_dir` as `work_dirs`. 
And this PR is almost done.

**Comment by zhouzaida** [comment]
> @zhouzaida I wrote a note in document. Recommend not to set `work_dir` as `work_dirs`. And this PR is almost done.

OK

File: mmengine/visualization/vis_backend.py [hunk 6]
```diff
@@ -1130,3 +1131,183 @@ def close(self) -> None:
...
+
+        Or the `log_params` method of dvclive will raise an error.
+        """
+
+        if isinstance(value, (dict, Config, ConfigDict)):
+            return {k: self._to_dvc_paramlike(v) for k, v in value.items()}
+        elif isinstance(value, tuple):
+            return [self._to_dvc_paramlike(item) for item in value]
+        elif isinstance(value, list):
+            return [self._to_dvc_paramlike(item) for item in value]
```
Comment (zhouzaida) [inline mmengine/visualization/vis_backend.py:1307]:
```suggestion
        elif isinstance(value, (tuple, list)):
            return [self._to_dvc_paramlike(item) for item in value]
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**